zcash_keys/src/keys.rs

```diff
@@ -785,48 +881,262 @@ impl UnifiedFullViewingKey {
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Constructs a unified incoming viewing key from a parsed unified encoding.
+    pub fn from_uivk(uivk: &Uivk) -> Result<Self, UnifiedError> {
+        #[cfg(feature = "orchard")]
+        let mut orchard = None;
+        #[cfg(feature = "sapling")]
+        let mut sapling = None;
+        #[cfg(feature = "transparent-inputs")]
+        let mut transparent = None;
+
+        // We can use as-parsed order here for efficiency, because we're breaking out the
+        // receivers we support from the unknown receivers.
+        let unknown = uivk
+            .items_as_parsed()
+            .iter()
+            .filter_map(|receiver| match receiver {
+                #[cfg(feature = "orchard")]
+                unified::Ivk::Orchard(data) => {
+                    Option::from(orchard::keys::IncomingViewingKey::from_bytes(data))
+                        .ok_or(UnifiedError::InvalidShieldedKey(ShieldedProtocol::Orchard))
+                        .map(|addr| {
+                            orchard = Some(addr);
+                            None
+                        })
+                        .transpose()
+                }
+                #[cfg(not(feature = "orchard"))]
+                unified::Ivk::Orchard(data) => Some(Ok::<_, UnifiedError>((
+                    u32::from(unified::Typecode::Orchard),
+                    data.to_vec(),
+                ))),
+                #[cfg(feature = "sapling")]
+                unified::Ivk::Sapling(data) => {
+                    Option::from(::sapling::zip32::IncomingViewingKey::from_bytes(data))
+                        .ok_or(UnifiedError::InvalidShieldedKey(ShieldedProtocol::Sapling))
+                        .map(|pa| {
+                            sapling = Some(pa);
+                            None
+                        })
+                        .transpose()
+                }
+                #[cfg(not(feature = "sapling"))]
+                unified::Ivk::Sapling(data) => Some(Ok::<_, UnifiedError>((
+                    u32::from(unified::Typecode::Sapling),
+                    data.to_vec(),
+                ))),
+                #[cfg(feature = "transparent-inputs")]
+                unified::Ivk::P2pkh(data) => legacy::ExternalIvk::deserialize(data)
+                    .map_err(UnifiedError::InvalidTransparentKey)
+                    .map(|tivk| {
+                        transparent = Some(tivk);
+                        None
+                    })
+                    .transpose(),
+                #[cfg(not(feature = "transparent-inputs"))]
+                unified::Ivk::P2pkh(data) => Some(Ok::<_, UnifiedError>((
+                    u32::from(unified::Typecode::P2pkh),
+                    data.to_vec(),
+                ))),
+                unified::Ivk::Unknown { typecode, data } => Some(Ok((*typecode, data.clone()))),
+            })
+            .collect::<Result<_, _>>()?;
```

Comment:
```suggestion

        let mut unknown = vec![];

        // We can use as-parsed order here for efficiency, because we're breaking out the
        // receivers we support from the unknown receivers.
        for receiver in uivk.items_as_parsed() {
            match receiver {
                unified::Ivk::Orchard(data) => {
                    #[cfg(feature = "orchard")]
                    {
                        orchard = Some(
                            Option::from(orchard::keys::IncomingViewingKey::from_bytes(data))
                                .ok_or(UnifiedError::InvalidShieldedKey(
                                    ShieldedProtocol::Orchard,
                                ))?,
                        );
                    }

                    #[cfg(feature = "orchard")]
                    unknown.push((u32::from(unified::Typecode::Orchard), data.to_vec()));
                }
                unified::Ivk::Sapling(data) => {
                    #[cfg(feature = "sapling")]
                    {
                        sapling = Some(
                            Option::from(::sapling::zip32::IncomingViewingKey::from_bytes(data))
                                .ok_or(UnifiedError::InvalidShieldedKey(
                                    ShieldedProtocol::Sapling,
                                ))?,
                        );
                    }

                    #[cfg(not(feature = "sapling"))]
                    unknown.push((u32::from(unified::Typecode::Sapling), data.to_vec()));
                }
                unified::Ivk::P2pkh(data) => {
                    #[cfg(feature = "transparent-inputs")]
                    {
                        transparent = Some(
                            legacy::ExternalIvk::deserialize(data)
                                .map_err(UnifiedError::InvalidTransparentKey)?,
                        );
                    }

                    #[cfg(not(feature = "transparent-inputs"))]
                    unknown.push((u32::from(unified::Typecode::P2pkh), data.to_vec()));
                }
                unified::Ivk::Unknown { typecode, data } => {
                    unknown.push((*typecode, data.clone()));
                }
            }
        }
```

The style that this was done in for the ufvk doesn't really make much sense; the direct mutation approach is much clearer. I had made a similar change to the ufvk handling in #1135 so we shouldn't reproduce the ugly approach here.